common/crypto/platform/openssl/crypto.cpp

```diff
@@ -148,8 +160,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
+ExpectedPrivateKey PrivateKey::LoadFromHSM(const Args &args) {
+	log::Debug("Loading the private key from HSM");
+
+	auto default_provider = unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)>(
+		OSSL_PROVIDER_load(NULL, "default"), OSSL_PROVIDER_unload);
+	if (default_provider.get() == nullptr) {
+		return expected::unexpected(
+			MakeError(SetupError, "default provider load error: " + GetOpenSSLErrorMessage()));
+	}
+
+	auto hsm_provider = unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)>(
+		OSSL_PROVIDER_load(NULL, args.ssl_engine.c_str()), OSSL_PROVIDER_unload);
+	if (hsm_provider.get() == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError, args.ssl_engine + " provider load error: " + GetOpenSSLErrorMessage()));
+	}
+
+	int ret {OSSL_PROVIDER_available(NULL, args.ssl_engine.c_str())};
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError, args.ssl_engine + " provider not available: " + GetOpenSSLErrorMessage()));
+	}
+
+	log::Trace("Loading key: " + args.private_key_path);
+	auto ctx = unique_ptr<OSSL_STORE_CTX, int (*)(OSSL_STORE_CTX *)>(
+		OSSL_STORE_open(args.private_key_path.c_str(), NULL, NULL, NULL, NULL), OSSL_STORE_close);
+
+	if (ctx == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"OSSL_STORE_OPEN: Failed to load the private key from the hardware security module: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	// Go through all objects in the context till we find the first private key
+	while (not OSSL_STORE_eof(ctx.get())) {
+		OSSL_STORE_INFO *info = OSSL_STORE_load(ctx.get());
+
+		if (info == NULL) {
+			return expected::unexpected(MakeError(
+				SetupError,
+				"Failed to read the private key information the hardware security module: "
+					+ GetOpenSSLErrorMessage()));
+		}
+
+		const int type_info {OSSL_STORE_INFO_get_type(info)};
+		switch (type_info) {
+		case OSSL_STORE_INFO_PKEY: {
+			// NOTE: get1 creates a duplicate of the pkey from the info, which can be
+			// used after the info ctx is destroyed
+			auto private_key = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>(
+				OSSL_STORE_INFO_get1_PKEY(info), pkey_free_func);
+			if (private_key == nullptr) {
+				return expected::unexpected(MakeError(
+					SetupError,
+					"Failed to load the private key from the hardware security module: "
+						+ GetOpenSSLErrorMessage()));
+			}
+			log::Debug("Successfully loaded the private key");
+			return PrivateKey(
+				std::move(private_key), std::move(default_provider), std::move(hsm_provider));
+		}
+		default:
+			const string info_type_string = OSSL_STORE_INFO_type_string(type_info);
+			return expected::unexpected(MakeError(
+				SetupError,
+				"Unhandled OpenSSL type: expected PrivateKey, got: " + info_type_string));
+			break;
+		}
+	}
+
+	return expected::unexpected(MakeError(
+		SetupError,
+		"Failed to load the private key from the hardware security module: "
+			+ GetOpenSSLErrorMessage()));
+}
```

Comment:
It is really hard to follow. I am glad we have a new OpenSSL expert in the team :sweat_smile: 

I browsed the man pages and your code makes sense, but I am not able to provide any better feedback than that.

Reply:
It's a shame, I have been trying to make this as readable as possible. I could perhaps comment it out more?